### PR TITLE
2842 improve data migration code base

### PR DESCRIPTION
**Pull request by sambodeme**
## Description

- Added logs
- Added tests to core functions
- Removed fake/default data
- Removed scan and s3/load for data migration workbook objects
- Refactored/Cleaned/Improved Code
#2842 

## For Testing
Run any migration command specified in the README

## PR checklist: submitters

- [ ]   Link to an issue if possible. If there’s no issue, describe what your branch does. Even if there is an issue, a brief description in the PR is still useful.
- [ ]   List any special steps reviewers have to follow to test the PR. For example, adding a local environment variable, creating a local test file, etc.
- [ ]   For extra credit, submit a screen recording like [this one](https://github.com/GSA-TTS/FAC/pull/1821).
- [ ]   Make sure you’ve merged `main` into your branch shortly before creating the PR. (You should also be merging `main` into your branch regularly during development.)
- [ ]   Make sure you’ve accounted for any migrations. When you’re about to create the PR, bring up the application locally and then run `git status | grep migrations`. If there are any results, you probably need to add them to the branch for the PR. Your PR should have only **one** new migration file for each of the component apps, except in rare circumstances; you may need to delete some and re-run `python manage.py makemigrations` to reduce the number to one. (Also, unless in exceptional circumstances, your PR should not delete any migration files.)
- [ ]   Make sure that whatever feature you’re adding has tests that cover the feature. This includes test coverage to make sure that the previous workflow still works, if applicable.
- [ ]   Make sure the full-submission.cy.js [Cypress test](https://github.com/GSA-TTS/FAC/blob/main/docs/testing.md#end-to-end-testing) passes, if applicable.
- [ ]   Do manual testing locally. Our tests are not good enough yet to allow us to skip this step. If that’s not applicable for some reason, check this box.
- [ ]   Verify that no Git surgery was necessary, or, if it was necessary at any point, repeat the testing after it’s finished.
- [ ]   Once a PR is merged, keep an eye on it until it’s deployed to dev, and do enough testing on dev to verify that it deployed successfully, the feature works as expected, and the happy path for the broad feature area (such as submission) still works.

## PR checklist: reviewers

- [ ]   Pull the branch to your local environment and run `make docker-clean; make docker-first-run && docker compose up`; then run `docker compose exec web /bin/bash -c "python manage.py test"`
- [ ]   Manually test out the changes locally, or check this box to verify that it wasn’t applicable in this case.
- [ ]   Check that the PR has appropriate tests. Look out for changes in HTML/JS/JSON Schema logic that may need to be captured in Python tests even though the logic isn’t in Python.
- [ ]   Verify that no Git surgery is necessary at any point (such as during a merge party), or, if it was, repeat the testing after it’s finished.

The larger the PR, the stricter we should be about these points.

### CODE REVIEW

File: backend/census_historical_migration/transforms/xform_string_to_int.py [hunk 1]
```diff
@@ -1,12 +1,9 @@
 def string_to_int(value):
     """Converts a string to an integer."""
-    if isinstance(value, int):
-        return value
+
     if not isinstance(value, str):
         raise ValueError(f"Expected string, got {type(value).__name__}")
-
     value = value.strip()
-
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/transforms/xform_string_to_int.py:9]: My pref would be to restore these

File: backend/census_historical_migration/transforms/xform_string_to_string.py [hunk 2]
```diff
@@ -1,13 +1,11 @@
 def string_to_string(value):
     """
     Converts a string to a trimmed string. Returns an empty string if the input
-    is empty or 'nan'."""
+    is None."""
     if value is None:
         return ""
     if not isinstance(value, str):
         raise ValueError(f"Expected string, got {type(value).__name__}")
     trimmed_value = value.strip()
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/transforms/xform_string_to_string.py:9]:
```suggestion
    if value is None:
        return ""
 
    if not isinstance(value, str):
        raise ValueError(f"Expected string, got {type(value).__name__}")
```

File: backend/census_historical_migration/transforms/xform_string_to_string.py [hunk 3]
```diff
@@ -1,13 +1,11 @@
...
         return ""
     if not isinstance(value, str):
         raise ValueError(f"Expected string, got {type(value).__name__}")
     trimmed_value = value.strip()
-    # FIXME-MSHD: When some CSV files are loaded
-    # to Postgres DB, empty string are being converted into 'nan'
-    # This is a temporary fix to handle the issue, we need to investigate this further.
-    return "" if trimmed_value in ["nan"] else trimmed_value
+
+    return trimmed_value
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/transforms/xform_string_to_string.py:11]:
```suggestion
    return value.strip()
```

File: backend/census_historical_migration/workbooklib/additional_eins.py [hunk 4]
```diff
@@ -53,15 +62,15 @@ def generate_additional_eins(audit_header, outfile):
...
     addl_eins = _get_eins(audit_header.DBKEY, audit_header.AUDITYEAR)
     map_simple_columns(wb, mappings, addl_eins)
     wb.save(outfile)
 
     table = generate_dissemination_test_table(
         audit_header, "additional_eins", mappings, addl_eins
     )
-    table["singletons"]["auditee_uei"] = audit_header.UEI
+    table["singletons"]["auditee_uei"] = uei
     return (wb, table)
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/workbooklib/additional_eins.py:76]:
```suggestion
    table["singletons"]["auditee_uei"] = uei

    return (wb, table)
```

File: backend/census_historical_migration/workbooklib/additional_ueis.py [hunk 5]
```diff
@@ -34,15 +34,15 @@ def generate_additional_ueis(audit_header, outfile):
...
     additional_ueis = _get_ueis(audit_header.DBKEY, audit_header.AUDITYEAR)
     map_simple_columns(wb, mappings, additional_ueis)
     wb.save(outfile)
 
     table = generate_dissemination_test_table(
         audit_header, "additional_ueis", mappings, additional_ueis
     )
-    table["singletons"]["auditee_uei"] = audit_header.UEI
+    table["singletons"]["auditee_uei"] = uei
     return (wb, table)
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/workbooklib/additional_ueis.py:48]:
```suggestion
    table["singletons"]["auditee_uei"] = uei

    return (wb, table)
```

File: backend/census_historical_migration/workbooklib/corrective_action_plan.py [hunk 6]
```diff
@@ -41,16 +42,17 @@ def generate_corrective_action_plan(audit_header, outfile):
...
     captexts = _get_cap_text(audit_header.DBKEY, audit_header.AUDITYEAR)
     map_simple_columns(wb, mappings, captexts)
     wb.save(outfile)
 
     table = generate_dissemination_test_table(
         audit_header, "corrective_action_plans", mappings, captexts
     )
-    table["singletons"]["auditee_uei"] = audit_header.UEI
+    table["singletons"]["auditee_uei"] = uei
     return (wb, table)
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/workbooklib/corrective_action_plan.py:58]:
```suggestion
    table["singletons"]["auditee_uei"] = uei

    return (wb, table)
```

File: backend/census_historical_migration/workbooklib/federal_awards.py [hunk 7]
```diff
@@ -216,14 +222,16 @@ def _xform_populate_default_loan_balance(loans_at_end, audits):
     If the audit's LOANS attribute is "Y" and the loan balance is empty,
     it fills in a default value indicating that no loan balance was provided."""
     for ndx, audit in zip(range(len(audits)), audits, loans_at_end):
-        if audit.LOANS == "Y":
-            if audit.LOANBALANCE is None:
+        loan = string_to_string(audit.LOANS).upper()
+        balance = string_to_string(audit.LOANBALANCE)
+        if loan == "Y":
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/workbooklib/federal_awards.py:227]:
```suggestion
        balance = string_to_string(audit.LOANBALANCE)

        if loan == "Y":
```

File: backend/census_historical_migration/workbooklib/federal_awards.py [hunk 8]
```diff
@@ -216,14 +222,16 @@ def _xform_populate_default_loan_balance(loans_at_end, audits):
...
+            if not balance:
                 loans_at_end[
                     ndx
                 ] = 1  # FIXME - MSHD: This value requires team approval.
                 # There are cases (dbkeys 148665/150450) with balance = -1.0, how do we handle this?
         else:
-            if audit.LOANBALANCE is not None:
+            if not balance:
                 loans_at_end[ndx] = ""
     return loans_at_end
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/workbooklib/federal_awards.py:236]:
```suggestion
                loans_at_end[ndx] = ""

    return loans_at_end
```

File: backend/census_historical_migration/management/commands/historic_data_migrator.py [hunk 9]
```diff
@@ -53,13 +58,15 @@ def initiate_migration(self, dbkeys_str, years_str):
                     continue
 
                 run_end_to_end(user, audit_header, result)
-                logger.info(result)
+                result_log[(year, dbkey)] = result
+                total_count += 1
+                has_failed = len(result["errors"]) > 0
+                if has_failed:
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/management/commands/historic_data_migrator.py:64]:
```suggestion
                total_count += 1

                has_failed = len(result["errors"]) > 0
                if has_failed:
```

**Review by phildominguez-gsa — CHANGES_REQUESTED** [review]
Looking good, mostly whitespace suggestions for readability. Take/leave/resolve as you see fit!

File: backend/census_historical_migration/sac_general_lib/general_information.py [hunk 10]
```diff
@@ -123,23 +129,48 @@ def _census_audit_type(s):
     return AUDIT_TYPE_DICT[s]
 
 
-def _xform_country(general_information):
+def xform_country(general_information, audit_header):
```
[the remark this answers is not in the document]
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/general_information.py:132]: We have many records with the `CPACOUNTRY` field empty, even though the `CPASTATE` exists and matches U.S. state 
two-digit abbreviations. The old approach incorrectly defaulted all of these records to non-USA. That is wrong. 
The update here is to address that approach. In cases of `non-USA`, if any, the report will fail migration, 
and we can better assess and make the right transformation at the time.

File: backend/census_historical_migration/sac_general_lib/general_information.py [hunk 11]
```diff
@@ -21,15 +25,16 @@
...
+    For context, see ticket #2912.
+    """
     mappings = {
         r"institution\s+of\s+higher\s+education": "higher-ed",
         r"non-?profit": "non-profit",
         r"local\s+government": "local",
         r"state": "state",
         r"unknown": "unknown",
-        r"": "none",
-        # r"": "tribal"   FIXME-MSHD: what is being used for tribal in Census table?
```
Comment (gsa-suk) [inline backend/census_historical_migration/sac_general_lib/general_information.py:32]: "indian tribe or tribal organization" is the ENTITY_TYPE in Census.

File: backend/census_historical_migration/historic_data_loader.py [hunk 12]
```diff
@@ -33,21 +33,23 @@ def load_historic_data_for_year(audit_year, page_size, pages):
...
-            if error_count > 5:
-                break
+
+    print_results(result_log, error_count, total_count)
+
+
+def print_results(result_log, error_count, total_count):
+    """Prints the results of the migration"""
 
     print("********* Loader Summary ***************")
```
Comment (jadudm) [inline backend/census_historical_migration/historic_data_loader.py:54]: Flip `print` to `logger.log`.

File: backend/census_historical_migration/sac_general_lib/report_id_generator.py [hunk 13]
```diff
@@ -1,15 +1,20 @@
...
 from ..sac_general_lib.utils import (
     xform_census_date_to_datetime,
 )
 
 
 def xform_dbkey_to_report_id(audit_header):
+    """Constructs the report ID from the DBKEY.
+    The report ID is constructed as follows:
+    YYYY-MM-CENSUS-DBKEY."""
     # month = audit_header.fyenddate.split('-')[1]
```
Comment (jadudm) [inline backend/census_historical_migration/sac_general_lib/report_id_generator.py:11]: We should remove/clean up any comments that are lingering at this point, or make sure they're accurate.

**Review by jadudm — APPROVED** [review]
LGTM. Code review w/ MSHD, SK, PD.

File: backend/census_historical_migration/historic_data_loader.py (hunk 12 above)
Comment (sambodeme) [reply backend/census_historical_migration/historic_data_loader.py:54]: Added this as task in https://github.com/GSA-TTS/FAC/issues/2837

File: backend/census_historical_migration/management/commands/historic_data_migrator.py [hunk 14]
```diff
@@ -43,6 +46,8 @@ def initiate_migration(self, dbkeys_str, years_str):
             logger.info(
                 f"Generating test reports for DBKEYS: {dbkeys_str} and YEARS: {years_str}"
             )
+            result_log = {}
+            total_count = error_count = 0
             for dbkey, year in zip(dbkeys, years):
```
[the remark this answers is not in the document]
Comment (sambodeme) [reply backend/census_historical_migration/management/commands/historic_data_migrator.py:51]: Let's defer this to dev huddle.

File: backend/census_historical_migration/management/commands/historic_data_migrator.py (hunk 9 above)
Comment (sambodeme) [reply backend/census_historical_migration/management/commands/historic_data_migrator.py:64]: same as previous

File: backend/census_historical_migration/sac_general_lib/report_id_generator.py (hunk 13 above)
Comment (sambodeme) [reply backend/census_historical_migration/sac_general_lib/report_id_generator.py:11]: Captured in https://github.com/GSA-TTS/FAC/issues/2837

**Review by phildominguez-gsa — APPROVED** [review]

**Pull request merged**